File: GFG/BT/AreCousins.java
import java.lang.*;
import java.util.*;

// #!/usr/bin/python -O

// #include <stdio.h>
// #include <stdlib.h>
// #include <string.h>
// #include<stdbool.h>
// #include<limits.h>

// #include<iostream>
// #include<algorithm>
// #include<string>
// #include<vector>
//using namespace std;

/*

# Author			: @RAJ009F
# Topic or Type 	: GFG/BT
# Problem Statement	:  
# Description		: 
# Complexity		: 
=======================
#steps:
=----------------


#sample output
----------------------

=======================
*/
class Node
	{
		Node l;
		Node r;
		int data;
		
		Node(int data)
			{
				this.data = data;
				this.l = null;
				this.r =  null;
			
			}
	
	
	}
	
class Util
{
	Node parent;

	int level;
	
	
	
}	
	
	class AreCousins
	{
		Node root;
		
		
		public  void getLevelandParent(Node h, Node target,  int level,Util u)
		{
			if(h == null)
				return;
			if(h.l==target || h.r==target)
			{
				u.level = level+1;
				u.parent = h;
			}
			
			getLevelandParent(h.l, target, level+1, u );
			getLevelandParent(h.r, target, level+1, u);
			
		}
		
		public boolean areCousins(Node h, Node node1, Node node2)
		{
			if(h==null)
				return false;
			
			Util u =  new Util();
			
			getLevelandParent(h, node1, 0,u);
			Node parent = u.parent;
			int level = u.level;
			
			Util v=  new Util();
			getLevelandParent(h, node1, 0, v);
			
			System.out.println(parent.data + " "+ level);
			System.out.println(v.parent.data + " "+ v.level);
			return (parent==v.parent && v.level==level);
			
		}
		
		public static void main(String args[])
		{
		
			AreCousins bt =  new AreCousins();
			bt.root =  new Node(1);
				bt.root.l = new Node(2);
				bt.root.r = new Node(3);
				bt.root.l.l = new Node(4);
				bt.root.l.r = new Node(5);
				bt.root.r.l =  new Node(6);
				bt.root.r.r = new Node(7);
				
				System.out.println(bt.areCousins(bt.root, bt.root.r.r , bt.root.l.l));	
				System.out.println(bt.areCousins(bt.root, bt.root.r.r , bt.root.l));	
		
			
			
		}
	
	
	
	}

File: GFG/BT/AreCousins.c
// import java.lang.*;
// import java.util.*;

// #!/usr/bin/python -O

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include<stdbool.h>
#include<limits.h>

// #include<iostream>
// #include<algorithm>
// #include<string>
// #include<vector>
//using namespace std;

/*

# Author			: @RAJ009F
# Topic or Type 	: GFG/BT
# Problem Statement	:Check whether two nodes are cpousins
# Description		: 
# Complexity		: 
=======================
#steps:
=----------------


#sample output
----------------------

=======================
*/

struct Node
{
	struct Node *l;
	struct Node  *r;
	int data;
};


struct Node *newNode(int data)
		{
			struct Node *node =  (struct Node *) malloc(sizeof(struct Node ));
			node->data =data;
			node->r = node->l = NULL;
			return node;		
		}


void getLevelandParent(struct Node *root, struct Node *target, int l, int *level, struct Node **parent)
{
	if(!root)
		return ;
	if(root->l == target||root->r == target)
	{
		*level = l+1;
		*parent = root;
		
	}
getLevelandParent (root->l, target, l+1, level, parent);


getLevelandParent (root->r, target, l+1, level, parent);
	
	
	
	
}
bool areCousins(struct Node *root, struct Node *node1, struct Node *node2)	
{
	
	struct Node *parent1=NULL; struct Node *parent2=NULL;
	int level1=0, level2=0;
	
	if(!root)
		return false;
	getLevelandParent(root, node1,0, &level1, &parent1);
	
	getLevelandParent(root, node2, 0, &level2, &parent2);
	
	return (level1==level2 && parent1!=parent2);
	
	
}	
		
int main()
{
	struct Node *root = newNode(1);
		root->l = newNode(2);
		root->l->l = newNode(4);
		root->l->r = newNode(5);
		
		root->r = newNode(3);
		root->r->l = newNode(6);
		root->r->l->r = newNode(8);
		root->r->r = newNode(7);
		
		areCousins(root, root->l, root->r->r)?printf("true"):printf("false");
		areCousins(root, root->r->r, root->l->l )? printf("true"):printf("false");
		
}		